### PR TITLE
Also read reweighting info etc.

### PR DESCRIPTION
**Pull request by iamholger**
LHEInit and LHEProcIfno are some additional things of the lhef3 standard (according to Stefan Prestel).
They re so simple, I just left them as dicts.

The LHEEvent now also has attributes and weights. The attributes are supposedly abused for reweighting information.

In order to not mess up things to badly, I made a copy of your readLHE and extended it to also fill the weights and attributes properties.

Finally, readNumEvents returns the number of events in the file.

### CODE REVIEW

File: pylhe/__init__.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
 import os
```
Comment (lukasheinrich) [inline pylhe/__init__.py]: add logging imports and set up

```python
import logging
log = logging.getLogger(__name__)
```

File: pylhe/__init__.py [hunk 2]
```diff
@@ -31,27 +33,94 @@ def __init__(self, **kwargs):
...
+        if element.tag == "initrwgt":
+            initDict["weightgroup"]={}
+            for child in element:
+                # Find all weightgroups
+                if child.tag=="weightgroup" and child.attrib!={}:
+                    try:
+                        wg_type = child.attrib["type"]
+                    except Exception, e:
+                        print "weightgroup must have attribute 'type'"
+                        sys.exit(1)
```
Comment (lukasheinrich) [inline pylhe/__init__.py]: Is there a reason we need to exit hard here?  The raised exception should already make the progam halt.  Also the exception should probably we narrowed down to, I assume, `KeyError`. If we want to add a meaningful message we can do

```
                 try:
                     wg_type = child.attrib["type"]
                  except KeyError:
                     log.exception("weightgroup must have attribute 'type'")
```

File: pylhe/__init__.py [hunk 3]
```diff
@@ -31,27 +33,94 @@ def __init__(self, **kwargs):
...
+                    _temp = {"attrib":child.attrib}
+                    _temp["weights"] = {}
+                    # Iterate over all weights in this weightgroup
+                    for w in child:
+                        if not w.tag=="weight": continue
+                        try:
+                            wg_id = w.attrib["id"]
+                        except Exception, e:
+                            print "weight must have attribute 'id'"
+                            sys.exit(1)
```
Comment (lukasheinrich) [inline pylhe/__init__.py]: same as comment above

File: pylhe/__init__.py [hunk 4]
```diff
@@ -60,11 +129,50 @@ def readLHE(thefile):
...
+            if sub.tag=="rwgt":
+                for r in sub:
+                    if r.tag=="wgt":
+                        eventdict["weights"][r.attrib["id"]] = float(r.text.strip())
+        # yield eventdict
+        yield LHEEvent(eventdict["eventinfo"], eventdict["particles"], eventdict["weights"], eventdict["attrib"])
+
+  except ET.ParseError:
+    print "WARNING. Parse Error."
+    return
```
Comment (lukasheinrich) [inline pylhe/__init__.py]: we should change this to `log.warning(..)` instead of print

**Comment by lukasheinrich** [comment]
Thanks @iamholger for the PR. I added some small comments, mainly abbout logging and exception handling otherwise this looks nice!

File: pylhe/__init__.py (hunk 2 above)
Comment (iamholger) [reply pylhe/__init__.py]: no, should have ben an exception in the first place, sorry

File: pylhe/__init__.py (hunk 4 above)
Comment (iamholger) [reply pylhe/__init__.py]: cool!

**Comment by iamholger** [comment]
Tahnks so much!

**Comment by lukasheinrich** [comment]
why did you close the PR? :)

**Comment by marinang** [comment]
These new features are really helpful actually to recover the weights from the lhefile. 
I was using Pythia python bindings to shower events from a lhefile but you cannot recover the weights with the current bindings ....

**Comment by lukasheinrich** [comment]
@iamholger do we want to revive this PR? @marinang if you already have a working version maybe you can push it upstream?

**Comment by marinang** [comment]
@lukasheinrich for sure I cannot push in this PR, let's wait a bit  if @iamholger  wants to revive it then I can give some suggestions.

**Comment by iamholger** [comment]
I think this is a good idea --- in principle we could
aggregate the hdf5 based scripts here, too.

Holger

**Comment by lukasheinrich** [comment]
@marinang have you added any patches to the original PR? maybe we can integrate them or merge @iamholger's PR first and add your patches on that new `master`

File: pylhe/__init__.py [hunk 5]
```diff
@@ -31,27 +33,94 @@ def __init__(self, **kwargs):
...
+            data = element.text.split('\n')[1:-1]
+            initDict["initInfo"]=LHEInit.fromstring(data[0])
+            initDict["procInfo"]=[LHEProcInfo.fromstring(d) for d in data[1:]]
+        if element.tag == "initrwgt":
+            initDict["weightgroup"]={}
+            for child in element:
+                # Find all weightgroups
+                if child.tag=="weightgroup" and child.attrib!={}:
+                    try:
+                        wg_type = child.attrib["type"]
```
Comment (marinang) [inline pylhe/__init__.py]: I have only changed here the tag, in my lhefiles produces by MadGraph it's `name` and not `weightgroup`.

**Comment by marinang** [comment]
@lukasheinrich So I think I could be merged as it is now and I can just implement small patches in another PR.

**Comment by lukasheinrich** [comment]
hi @marinang I merged this... feel free to add your patch.  thanks @iamholger for the PR :)

**Pull request merged**